enterprise/tailnet/pgcoord.go

```diff
@@ -106,30 +109,105 @@ func NewPGCoord(ctx context.Context, logger slog.Logger, ps pubsub.Pubsub, store
-	cCh := make(chan *connIO)
+	cCh := make(chan agpl.Queue)
+	sCh := make(chan subscribe)
 	// signals when first heartbeat has been sent, so it's safe to start binding.
 	fHB := make(chan struct{})
 
 	c := &pgCoord{
-		ctx:            ctx,
-		cancel:         cancel,
-		logger:         logger,
-		pubsub:         ps,
-		store:          store,
-		binder:         newBinder(ctx, logger, id, store, bCh, fHB),
-		bindings:       bCh,
-		newConnections: cCh,
-		id:             id,
-		querier:        newQuerier(ctx, logger, ps, store, id, cCh, numQuerierWorkers, fHB),
-		closed:         make(chan struct{}),
+		ctx:              ctx,
+		cancel:           cancel,
+		logger:           logger,
+		pubsub:           ps,
+		store:            store,
+		binder:           newBinder(ctx, logger, id, store, bCh, fHB),
+		bindings:         bCh,
+		newConnections:   cCh,
+		subscriber:       newSubscriber(ctx, logger, id, store, sCh, fHB),
+		newSubscriptions: sCh,
+		id:               id,
+		querier:          newQuerier(ctx, logger, id, ps, store, id, cCh, numQuerierWorkers, fHB),
+		closed:           make(chan struct{}),
 	}
 	logger.Info(ctx, "starting coordinator")
 	return c, nil
 }
 
 func (c *pgCoord) ServeMultiAgent(id uuid.UUID) agpl.MultiAgentConn {
-	_, _ = c, id
-	panic("not implemented") // TODO: Implement
+	ma := (&agpl.MultiAgent{
+		ID:                id,
+		AgentIsLegacyFunc: func(agentID uuid.UUID) bool { return true },
+		OnSubscribe: func(enq agpl.Queue, agent uuid.UUID) (*agpl.Node, error) {
+			err := c.addSubscription(enq, agent)
+			return c.Node(agent), err
+		},
+		OnUnsubscribe: c.removeSubscription,
+		OnNodeUpdate: func(id uuid.UUID, node *agpl.Node) error {
+			return sendCtx(c.ctx, c.bindings, binding{
+				bKey: bKey{id, agpl.QueueKindClient},
+				node: node,
+			})
+		},
+		OnRemove: func(enq agpl.Queue) {
+			b := binding{
+				bKey: bKey{
+					id:   enq.UniqueID(),
+					kind: enq.Kind(),
+				},
+			}
+			if err := sendCtx(c.ctx, c.bindings, b); err != nil {
+				c.logger.Debug(c.ctx, "parent context expired while withdrawing binding", slog.Error(err))
+			}
+			if err := sendCtx(c.ctx, c.newSubscriptions, subscribe{
+				sKey:   sKey{clientID: id},
+				active: false,
+			}); err != nil {
+				c.logger.Debug(c.ctx, "parent context expired while withdrawing subscriptions", slog.Error(err))
+			}
+			c.querier.cleanupConn(enq)
+		},
+	}).Init()
+
+	if err := sendCtx(c.ctx, c.newConnections, agpl.Queue(ma)); err != nil {
+		// If we can't successfully send the multiagent, that means the
+		// coordinator is shutting down. In this case, just return a closed
+		// multiagent.
+		ma.CoordinatorClose()
+	}
+
+	return ma
+}
+
+func (c *pgCoord) addSubscription(q agpl.Queue, agentID uuid.UUID) error {
+	err := sendCtx(c.ctx, c.newSubscriptions, subscribe{
+		sKey: sKey{
+			clientID: q.UniqueID(),
+			agentID:  agentID,
+		},
+		active: true,
+	})
+	if err != nil {
+		return err
+	}
+
+	c.querier.newClientSubscription(q, agentID)
```

Comment:
architecturally, this is still mixing communication styles where pgCoord communicates with the querier both over channels and via callbacks.  This can cause weird ordering issues where the subscriptions are processed before the newConn, or subscriptions are added after the connection is cleaned up.

It was initially designed with a channel, but if you feed strongly it should be callbacks, I think that would be OK --- but it should be one or the other.  That means removing the channel for agents and single-agent clients too.